File: framework/web/src/main/java/top/bettercode/simpleframework/web/resolver/CentConverter.java
package top.bettercode.simpleframework.web.resolver;

import java.util.Collections;
import java.util.Set;
import org.jetbrains.annotations.NotNull;
import org.springframework.core.convert.TypeDescriptor;
import org.springframework.core.convert.converter.ConditionalGenericConverter;
import org.springframework.util.StringUtils;
import top.bettercode.lang.util.MoneyUtil;

/**
 * 原字符串(1.2)×100转长整型(120)注解
 */
public class CentConverter implements ConditionalGenericConverter {

  @Override
  public boolean matches(@NotNull TypeDescriptor sourceType, TypeDescriptor targetType) {
    return targetType.hasAnnotation(Cent.class);
  }

  @Override
  public Set<ConvertiblePair> getConvertibleTypes() {
    return Collections.singleton(new ConvertiblePair(String.class, Long.class));
  }

  @Override
  public Object convert(Object object, @NotNull TypeDescriptor sourceType, @NotNull TypeDescriptor targetType) {
    if (!StringUtils.hasText((String) object)) {
      return null;
    }
    return MoneyUtil.toCent((String) object);
  }

}

File: web/src/main/kotlin/top/bettercode/summer/web/resolver/CentConverter.kt
package top.bettercode.summer.web.resolver

import org.springframework.core.convert.TypeDescriptor
import org.springframework.core.convert.converter.ConditionalGenericConverter
import org.springframework.core.convert.converter.GenericConverter
import org.springframework.util.StringUtils
import top.bettercode.summer.tools.lang.util.MoneyUtil.toCent

/**
 * 原字符串(1.2)×100转长整型(120)注解
 */
class CentConverter : ConditionalGenericConverter {
    override fun matches(sourceType: TypeDescriptor, targetType: TypeDescriptor): Boolean {
        return targetType.hasAnnotation(Cent::class.java)
    }

    override fun getConvertibleTypes(): Set<GenericConverter.ConvertiblePair> {
        return setOf(GenericConverter.ConvertiblePair(String::class.java, Long::class.java))
    }

    override fun convert(`object`: Any?, sourceType: TypeDescriptor, targetType: TypeDescriptor): Any? {
        return if (!StringUtils.hasText(`object` as String)) {
            null
        } else toCent(`object`)
    }
}
